File: multiprocessing_cavity_flow.py
import numpy
from matplotlib import pyplot, cm
import time
import multiprocessing
import os

nx = 100
ny = 100
nt = 1
nit = 100
c = 1
dx = 2 / (nx - 1)
dy = 2 / (ny - 1)
x = numpy.linspace(0, 2, nx)
y = numpy.linspace(0, 2, ny)
X, Y = numpy.meshgrid(x, y)

rho = 1
nu = .1
dt = .001

u = numpy.zeros((ny, nx))
v = numpy.zeros((ny, nx))
p = numpy.zeros((ny, nx))
b = numpy.zeros((ny, nx))


def build_up_b(b, rho, dt, u, v, dx, dy):

    b[1:-1, 1:-1] = (rho * (1 / dt *
                            ((u[1:-1, 2:] - u[1:-1, 0:-2]) /
                             (2 * dx) + (v[2:, 1:-1] - v[0:-2, 1:-1]) / (2 * dy)) -
                            ((u[1:-1, 2:] - u[1:-1, 0:-2]) / (2 * dx)) ** 2 -
                            2 * ((u[2:, 1:-1] - u[0:-2, 1:-1]) / (2 * dy) *
                                 (v[1:-1, 2:] - v[1:-1, 0:-2]) / (2 * dx)) -
                            ((v[2:, 1:-1] - v[0:-2, 1:-1]) / (2 * dy)) ** 2))

    return b


def pressure_poisson(p, dx, dy, b):
    pn = numpy.empty_like(p)
    pn = p.copy()

    for q in range(nit):
        pn = p.copy()
        p[1:-1, 1:-1] = (((pn[1:-1, 2:] + pn[1:-1, 0:-2]) * dy ** 2 +
                          (pn[2:, 1:-1] + pn[0:-2, 1:-1]) * dx ** 2) /
                         (2 * (dx ** 2 + dy ** 2)) -
                         dx ** 2 * dy ** 2 / (2 * (dx ** 2 + dy ** 2)) *
                         b[1:-1, 1:-1])

        p[:, -1] = p[:, -2]  # dp/dx = 0 at x = 2
        p[0, :] = p[1, :]  # dp/dy = 0 at y = 0
        p[:, 0] = p[:, 1]  # dp/dx = 0 at x = 0
        p[-1, :] = 0  # p = 0 at y = 2

    return p


def step_u(lx, ux, ly, uy,  un_step, vn_step, p_step, u_temp):
    ret = numpy.zeros_like(u_temp)
    for i in range(lx, ux):
        for j in range(ly, uy):
            ret[i, j] = (un_step[i, j] -
                         un_step[i, j] * dt / dx *
                         (un_step[i, j] - un_step[i, j - 1]) -
                         vn_step[i, j] * dt / dy *
                         (un_step[i, j] - un_step[i - 1, j]) -
                         dt / (2 * rho * dx) * (p_step[i, j + 1] - p_step[i, j - 1]) +
                         nu * (dt / dx ** 2 *
                         (un_step[i, j + 1] - 2 * un_step[i, j] + un_step[i, j - 1]) +
                          dt / dy ** 2 *
                          (un_step[i + 1, j] - 2 * un_step[i, j] + un_step[i - 1, j])))
    return ret


def step_v(lx, ux, ly, uy, un_step, vn_step, p_step, v_temp):
    ret = numpy.zeros_like(v_temp)
    for i in range(lx, ux):
        for j in range(ly, uy):
            ret[i, j] = (vn_step[i, j] -
                       un_step[i, j] * dt / dx *
                        (vn_step[i, j] - vn_step[i, j - 1]) -
                       vn_step[i, j] * dt / dy *
                       (vn_step[i, j] - vn_step[i - 1, j]) -
                       dt / (2 * rho * dy) * (p_step[i + 1, j] - p_step[i - 1, j]) +
                       nu * (dt / dx ** 2 *
                             (vn_step[i, j + 1] - 2 * vn_step[i, j] + vn_step[i, j - 1]) +
                             dt / dy ** 2 *
                             (vn_step[i + 1, j] - 2 * vn_step[i, j] + vn_step[i - 1, j])))
    return ret


def cavity_flow(nt, u, v, dt, dx, dy, p, rho, nu):
    un = numpy.empty_like(u)
    vn = numpy.empty_like(v)
    b = numpy.zeros((ny, nx))

    pool = multiprocessing.Pool(os.cpu_count())

    for n in range(nt):

        un = u.copy()
        vn = v.copy()

        b = build_up_b(b, rho, dt, u, v, dx, dy)
        p = pressure_poisson(p, dx, dy, b)
        row, col = u.shape

        res_u = pool.starmap(step_u, [(1, round(row / 2), 1, round(col / 2), un, vn, p, u),
                                    (1, round(row / 2), round(col / 2), col - 1, un, vn, p, u),
                                    (round(row / 2), row - 1, 1, round(col / 2), un, vn, p, u),
                                    (round(row / 2), row - 1, round(col / 2), col - 1, un, vn, p, u)])

        u = numpy.zeros_like(u)
        for r in res_u:
            u += r

        res_v = pool.starmap(step_v, [(1, round(row / 2), 1, round(col / 2), un, vn, p, v),
                              (1, round(row / 2), round(col / 2), col - 1, un, vn, p, v),
                              (round(row / 2), row - 1, 1, round(col / 2), un, vn, p, v),
                              (round(row / 2), row - 1, round(col / 2), col - 1, un, vn, p, v)])

        v = numpy.zeros_like(v)
        for r in res_v:
            v += r

        u[0, :] = 0
        u[:, 0] = 0
        u[:, -1] = 0
        u[-1, :] = 1  # set velocity on cavity lid equal to 1
        v[0, :] = 0
        v[-1, :] = 0
        v[:, 0] = 0
        v[:, -1] = 0

    return u, v, p


if __name__ == '__main__':

    p = numpy.zeros((ny, nx))
    b = numpy.zeros((ny, nx))
    nt = 100

    start = time.time()

    u, v, p = cavity_flow(nt, u, v, dt, dx, dy, p, rho, nu)

    end = time.time()

    print(end - start)

    # x = ""
    # tokens = x.split("] [")
    # for x in range(100):
    #     token = (tokens[x][1:-1]).split(" ")
    #     print(token)
    #     for y in range(100):
    #         if token[y] != "":
    #             u[x, y] = float(token[y])
    #
    # x2 = ""
    # tokens2 = x2.split("] [")
    # for x in range(100):
    #     token2 = (tokens2[x][1:-1]).split(" ")
    #     for y in range(100):
    #         if token2[y] != "":
    #             v[x, y] = float(token2[y])


    fig = pyplot.figure(figsize=(11, 7), dpi=100)
    # plotting the pressure field as a contour
    pyplot.contourf(X, Y, p, alpha=0.5, cmap=cm.viridis)
    pyplot.colorbar()
    # plotting the pressure field outlines
    pyplot.contour(X, Y, p, cmap=cm.viridis)
    # plotting velocity field
    pyplot.quiver(X[::2, ::2], Y[::2, ::2], u[::2, ::2], v[::2, ::2])
    pyplot.xlabel('X')
    pyplot.ylabel('Y')
    pyplot.show()

File: multiprocessing_cavity_flow.go
package main

import (
	"fmt"
	"math"
	"time"
)

func main() {
	const nt int = 100
	const nx int = 100
	const ny int = 100
	var dt float64 = 0.001
	var rho float64 = 1.0
	var nu float64 = 0.1
	var dx float64 = 2 / (float64(nx) - 1)
	var dy float64 = 2 / (float64(ny) - 1)
	var u = zeros(nx, ny)
	var v = zeros(nx, ny)
	var p = zeros(nx, ny)
	
	start := time.Now()
	cavity_flow(nt, u, v, dt, dx, dy, p, rho, nu)
	end := time.Now()
	fmt.Println(end.Sub(start))
}

// python numpy.zeros( ... )
func zeros(rows int, cols int) [][]float64 {
	a := make([][]float64, rows)
	for i := range a {
		a[i] = make([]float64, cols)
	}

	return a
}

func build_up_b(b [][]float64, rho float64, dt float64, u [][]float64, v [][]float64, dx float64, dy float64) [][]float64 {

	var rows = len(b)
	var cols = len(b[0])

	for i := 1; i < (rows - 1); i++ {
		for j := 1; j < (cols - 1); j++ {
			b[i][j] = (rho * (1/dt*
				((u[i][j+1]-u[i][j-1])/
					(2*dx)+(v[i+1][j]-v[i-1][j])/(2*dy)) -
				math.Pow(((u[i][j+1]-u[i][j-1])/(2*dx)), 2) -
				2*((u[i+1][j]-u[i-1][j])/(2*dy)*
					(v[i][j+1]-v[i][j-1])/(2*dx)) -
				math.Pow(((v[i+1][j]-v[i-1][j])/(2*dy)), 2)))
		}
	}

	return b
}

func copy(src [][]float64, dest [][]float64) {

	for i := range src {
		for j := range src[i] {
			dest[i][j] = src[i][j]
		}
	}
}

// python numpy.zeros_like( ... )
func copy_with_zeros(src [][]float64, dest [][]float64) {

	for i := range src {
		for j := range src[i] {
			dest[i][j] = 0
		}
	}
}


func add_matrix(base [][]float64, adding [][]float64){
	
	for i := range base{
		for j := range base[i]{
			if adding[i][j] == 0{

			} else {	
				base[i][j] = adding[i][j]
			} 
		}
	}
}

func pressure_poisson(p [][]float64, dx float64, dy float64, b [][]float64, niter int) [][]float64 {

	var rows = len(p)
	var cols = len(p[0])
	var pn = zeros(rows, cols)

	for iter := 0; iter < niter; iter++ {
		copy(p, pn)
		for i := 1; i < (rows - 1); i++ {
			for j := 1; j < (cols - 1); j++ {

				p[i][j] = (((pn[i][j+1]+pn[i][j-1])*math.Pow(dy, 2)+
					(pn[i+1][j]+pn[i-1][j])*math.Pow(dx, 2))/
					(2*(math.Pow(dx, 2)+math.Pow(dy, 2))) -
					math.Pow(dx, 2)*math.Pow(dy, 2)/(2*(math.Pow(dx, 2)+math.Pow(dy, 2)))*
						b[i][j])
			}
		}

		// python p[:,-1] = p[:,-2]
		// python p[:, 0] = p[:. 1]
		for x := 0; x < rows; x++ {
			p[x][cols-1] = p[x][cols-2]
			p[x][0] = p[x][1]
		}

		// python p[ 0,:] = p[1,:]
		// python p[-1,:] = 0
		for x := 0; x < cols; x++ {
			p[0][x] = p[1][x]
			p[rows-1][x] = 0
		}
	}

	return p
}


func step_u(lx int, ux int, ly int, uy int, un_step [][]float64, vn_step [][]float64, p_step [][]float64, u_temp [][]float64,
		dx float64, dy float64, dt float64, rho float64, nu float64, ch chan [][]float64) {
	

	var rows = len(u_temp)
	var cols = len(u_temp[0])
	
	var ret = zeros(rows, cols)
	for i := lx; i < ux; i++{
		for j := ly; j < uy; j++{
			ret[i][j] = (un_step[i][j] -
				un_step[i][j]*dt/dx*
				(un_step[i][j]-un_step[i][j-1]) -
				vn_step[i][j]*dt/dy*
				(un_step[i][j]-un_step[i-1][j]) -
				dt/(2*rho*dx)*(p_step[i][j+1]-p_step[i][j-1]) +
				nu*(dt/math.Pow(dx, 2)*
					(un_step[i][j+1]-2*un_step[i][j]+un_step[i][j-1])+
					dt/math.Pow(dy, 2)*
					(un_step[i+1][j]-2*un_step[i][j]+un_step[i-1][j])))
		}
	}

	ch <- ret
}



func step_v(lx int, ux int, ly int, uy int, un_step [][]float64, vn_step [][]float64, p_step [][]float64, v_temp [][]float64,
		dx float64, dy float64, dt float64, rho float64, nu float64, ch chan [][]float64) {
	
	var rows = len(v_temp)
	var cols = len(v_temp[0])
	
	var ret = zeros(rows, cols)
	for i := lx; i < ux; i++{
		for j := ly; j < uy; j++{
			ret[i][j] = (vn_step[i][j] -
					un_step[i][j]*dt/dx*
						(vn_step[i][j]-vn_step[i][j-1]) -
					vn_step[i][j]*dt/dy*
						(vn_step[i][j]-vn_step[i-1][j]) -
					dt/(2*rho*dy)*(p_step[i+1][j]-p_step[i-1][j]) +
					nu*(dt/math.Pow(dx, 2)*
						(vn_step[i][j+1]-2*vn_step[i][j]+vn_step[i][j-1])+
						dt/math.Pow(dy, 2)*
							(vn_step[i+1][j]-2*vn_step[i][j]+vn_step[i-1][j])))
		}
	}

	ch <- ret
}

func cavity_flow(nt int, u [][]float64, v [][]float64, dt float64, dx float64, dy float64, p [][]float64, rho float64, nu float64) ([][]float64, [][]float64, [][]float64) {

	var rows = len(u)
	var cols = len(u[0])

	var un = zeros(rows, cols)
	var vn = zeros(rows, cols)

	var b = zeros(rows, cols)

	for iter := 0; iter < nt; iter++ {
		
		copy(u, un)
		copy(v, vn)
		b = build_up_b(b, rho, dt, u, v, dx, dy)
		p = pressure_poisson(p, dx, dy, b, 50)

		ch_u := make(chan [][]float64, 4)
		ch_v := make(chan [][]float64, 4)
		
		go step_u(1, rows / 2, 1, cols / 2, un, vn, p, u, dx, dy, dt, rho, nu, ch_u)
		go step_u(1, rows / 2, cols / 2, cols - 1, un, vn, p, u, dx, dy, dt, rho, nu, ch_u)
		go step_u(rows / 2, rows - 1, 1, cols / 2, un, vn, p, u, dx, dy, dt, rho, nu, ch_u)
		go step_u(rows / 2, rows - 1, cols / 2, cols - 1, un, vn, p, u, dx, dy, dt, rho, nu, ch_u)
		
		go step_v(1, rows / 2, 1, cols / 2, un, vn, p, v, dx, dy, dt, rho, nu, ch_v)
		go step_v(1, rows / 2, cols / 2, cols - 1, un, vn, p, v, dx, dy, dt, rho, nu, ch_v)
		go step_v(rows / 2, rows - 1, 1, cols / 2, un, vn, p, v, dx, dy, dt, rho, nu, ch_v)
		go step_v(rows / 2, rows - 1, cols / 2, cols - 1, un, vn, p, v, dx, dy, dt, rho, nu, ch_v)
		
		//u_res := <- ch_u		
		//fmt.Println(u_res)
		//fmt.Println(<-ch_u)
		//fmt.Println(<-ch_u)
		//fmt.Println(<-ch_u)

		add_matrix(u, <-ch_u)
		add_matrix(u, <-ch_u)
		add_matrix(u, <-ch_u)
		add_matrix(u, <-ch_u)
		
		add_matrix(v, <-ch_v)
		add_matrix(v, <-ch_v)
		add_matrix(v, <-ch_v)
		add_matrix(v, <-ch_v)

		// python p[:,-1] = p[:,-2]
		// python p[:, 0] = p[:. 1]
		for x := 0; x < rows; x++ {
			u[x][cols-1] = 0
			u[x][0] = 0
			v[x][cols-1] = 0
			v[x][0] = 0
		}

		// python p[ 0,:] = p[1,:]
		// python p[-1,:] = 0
		for x := 0; x < cols; x++ {
			u[0][x] = 0
			u[rows-1][x] = 1
			v[0][x] = 0
			v[rows-1][x] = 0
		}
	}
	
	
	return u, v, p
}
